src/components/QueryOutput/QueryOutput.js

```diff
@@ -0,0 +1,1988 @@
+import errorMessages from '../../js/errorMessages'
+
+import {
+  onlyUnique,
+  formatElement,
+  makeEmptyArray,
+  getSupportedDisplayTypes,
+  getDefaultDisplayType,
+  isDisplayTypeValid,
+  getGroupBysFromPivotTable,
+  getGroupBysFromTable,
+  isTableType,
+  isChartType,
+  areAllColumnsHidden,
+  sortDataByDate,
+  dateSortFn,
+} from '../../js/Util.js'
+
+import {
+  isColumnNumberType,
+  isColumnStringType,
+  getNumberColumnIndices,
+  getDateColumnIndex,
+  getStringColumnIndices,
+  isColumnDateType,
+  isAggregation,
+} from './columnHelpers.js'
+
+import { sendSuggestion, runDrilldown } from '../../js/queryService'
+
+import './QueryOutput.scss'
+import { MONTH_NAMES } from '../../js/Constants'
+import { ReverseTranslation } from '../ReverseTranslation'
+import { getChartColorVars } from '../../theme/configureTheme'
+import { withTheme } from '../../theme'
+
+import './QueryOutput.scss'
+
+String.prototype.isUpperCase = function () {
+  return this.valueOf().toUpperCase() === this.valueOf()
+}
+
+String.prototype.toProperCase = function () {
+  return this.replace(/\w\S*/g, (txt) => {
+    if (txt.isUpperCase()) {
+      return txt
+    }
+    return txt.charAt(0).toUpperCase() + txt.substr(1).toLowerCase()
+  })
+}
+
+export class QueryOutput extends React.Component {
+  constructor(props) {
+    super(props)
+
+    this.COMPONENT_KEY = uuid()
+    this.QUERY_VALIDATION_KEY = uuid()
+
+    this.queryResponse = _cloneDeep(props.queryResponse)
+    this.queryID = _get(this.queryResponse, 'data.data.query_id')
+    this.interpretation = _get(this.queryResponse, 'data.data.interpretation')
+    this.tableID = uuid()
+    this.pivotTableID = uuid()
+    this.initialSupportedDisplayTypes = this.getCurrentSupportedDisplayTypes()
+
+    // Set initial config if needed
+    // If this config causes errors, it will be reset when the error occurs
+    if (
+      props.tableConfigs?.tableConfig &&
+      this.isTableConfigValid(props.tableConfigs?.tableConfig)
+    ) {
+      const { tableConfig, pivotTableConfig } = props.tableConfigs
+      this.tableConfig = _cloneDeep(tableConfig)
+      this.pivotTableConfig = _cloneDeep(pivotTableConfig)
+    }
+
+    // Set theme colors
+    const chartColors = getChartColorVars()
+    this.colorScale = scaleOrdinal().range(chartColors)
+
+    const columns = this.formatColumnsForTable(
+      props.queryResponse?.data?.data?.columns
+    )
+
+    // --------- generate data before mount --------
+    this.generateAllData()
+    // -------------------------------------------
+
+    // Supported display types may have changed after initial data generation
+    this.initialSupportedDisplayTypes = this.getCurrentSupportedDisplayTypes()
+
+    this.state = {
+      displayType: this.getDisplayTypeFromInitial(props),
+      supportedDisplayTypes: this.initialSupportedDisplayTypes,
+      columns,
+      tableFilters: [],
+      selectedSuggestion: props.selectedSuggestion,
```

Comment:
I think the idea behind this was so you could provide an initial selection, but didn't have to be responsible for keeping the state in the parent component. Mainly because QueryOutput can be used in isolation and I wanted there to be as little work as possible for the person using the component. I am definitely open to changing this though if it makes more sense to control it by props only

Reply:
Actually, I don't think we need this to be a prop at all. I just looked and there is no other component that passes in the selectedSuggestion prop. 

It might have been there before because at some point maybe the QueryOutput component was unmounted (maybe someone changed the page) and the current selection got reset, so I wanted a way to set it back to the last selection. We could keep it as an optional prop in this case if someone wanted to set it but instead of `selectedSuggestion` we could call it `defaultSelectedSuggestion`? 

What do you think?

Reply:
> I think the idea behind this was so you could provide an initial selection, but didn't have to be responsible for keeping the state in the parent component. Mainly because QueryOutput can be used in isolation and I wanted there to be as little work as possible for the person using the component. I am definitely open to changing this though if it makes more sense to control it by props only

Then, It makes sense. maybe just add some comments here and let others know why we are using props as initial state.